Add basic Spack installation resource

rkilpadi opened the pull request:
### Description of changes
Create a basic resource to install Spack in the default cluster user's home directory.

### Tests
Inspec tests to ensure that Spack directories have been created correctly and that user can successfully run `spack find` and `spack install`.
EC2 `./kitchen.ec2.sh platform-install verify spack -c 6`

### Checklist
- [ ] Make sure you are pointing to **the right branch** and add a label in the PR title (i.e. **2.x** vs **3.x**)
- [ ] Check all commits' messages are clear, describing what and why vs how.
- [ ] Make sure **to have added unit tests or integration tests** to cover the new/modified code.
- [ ] Check if documentation is impacted by this change.

Please review the [guidelines for contributing](../CONTRIBUTING.md) and [Pull Request Instructions](https://github.com/aws/aws-parallelcluster/wiki/Git-Pull-Request-Instructions).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

rkilpadi committed:
Increase size of default instance type

dreambeyondorange commented on cookbooks/aws-parallelcluster-platform/attributes/platform.rb:
```diff
@@ -13,6 +13,9 @@
 default['cluster']['nvidia']['driver_version'] = '470.182.03'
 
+# Spack
+default['cluster']['spack_root'] = "#{node['cluster']['base_dir']}/spack"
```
This should be on a shared storage location intended for the user.  I am thinking in the default user's home or something similar.  

['cluster']['base_dir'] is going to be in /opt which is intended to be an internal use location for pcluster

dreambeyondorange commented on cookbooks/aws-parallelcluster-platform/kitchen.platform-install.yml:
```diff
@@ -96,6 +96,22 @@ suites:
+      - recipe[aws-parallelcluster-tests::test_resource]
+    verifier:
+      controls:
+        - /tag:spack/
```
the convention right now is to use the tag in one of two ways, either with _install or _config based on whether it is baked into the AMI or run at boot.  If you're test doesn't use either of those yet, just use a name without `tag:`

dreambeyondorange commented on cookbooks/aws-parallelcluster-platform/kitchen.platform-install.yml:
```diff
@@ -96,6 +96,22 @@ suites:
+      dependencies:
+        - recipe:aws-parallelcluster-platform::directories
+        - resource:package_repos
+        - 'resource:package { "package_name": "git" }'
```
nice!

dreambeyondorange commented on cookbooks/aws-parallelcluster-platform/resources/spack_install.rb:
```diff
@@ -0,0 +1,37 @@
+default_action :install_spack
+
+action :install_spack do
+  spacksh_path = '/etc/profile.d/spack.sh'
```
this should probably be `spack_sh_path` or better `spack_script_path` to be more clear

dreambeyondorange commented on cookbooks/aws-parallelcluster-platform/resources/spack_install.rb:
```diff
@@ -0,0 +1,37 @@
+  end
+
+  execute 'change spack ownership to user' do
+    command "chown -R #{node['cluster']['cluster_user']}:#{node['cluster']['cluster_user']} #{node['cluster']['spack_root']}"
```
we probably don't want to be doing this in our pcluster /opt directories, another reason to change the root to a shared location for the user

dreambeyondorange commented on cookbooks/aws-parallelcluster-platform/test/controls/spack_spec.rb:
```diff
@@ -0,0 +1,36 @@
+    it { should be_grouped_into (node['cluster']['cluster_user']).to_s }
+  end
+
+  describe directory('/opt/parallelcluster/spack/share/spack') do
```
You should also use the variable here ['cluster']['spack_root']

dreambeyondorange commented on cookbooks/aws-parallelcluster-platform/test/controls/spack_spec.rb:
```diff
@@ -0,0 +1,36 @@
+control 'tag:spack_correctly_installed' do
+  only_if { instance.head_node? }
+
+  describe directory('/opt/parallelcluster/spack') do
```
You should probably use the variable here ['cluster']['spack_root']

dreambeyondorange commented on kitchen.ec2.yml:
```diff
@@ -22,7 +22,7 @@ driver:
     instance_metadata_tags: 'enabled'
   iam_profile_name: <%= ENV['KITCHEN_IAM_PROFILE'] %>
-  instance_type: <%= ENV['KITCHEN_INSTANCE_TYPE'] || 't2.micro' %>
+  instance_type: <%= ENV['KITCHEN_INSTANCE_TYPE'] || 'c5n.xlarge' %>
```
This unrelated change should be moved to a separate PR and it's also not mentioned in the PR description

Commit:
Remove scheduler-plugin code

Searched for: (case insensitive)
* "plugin" string,
* aws-parallelcluster-scheduler-plugin
* cluster-config-path

Signed-off-by: Enrico Usai <usai@amazon.com>

Commit:
Move third-party cookbooks into cookbooks folder

Berksfiles and CI/CD configuration files have been updated to use the new relative path.

Signed-off-by: Enrico Usai <usai@amazon.com>

Commit:
Explain code structure in the README file

Signed-off-by: Enrico Usai <usai@amazon.com>

Commit:
Skip mount_shared recipe tests on head node

These steps are executed in the compute nodes only.
Remove tag:config since this test requires some folders to be exported from the head node.

Signed-off-by: Enrico Usai <usai@amazon.com>

jdeamicis committed:
Allow override node package at cluster creation/update (#2300)

Add possibility to override node package at cluster creation and update
for development purposes. Only the head node will receive the overridden
node package at update time (we currently do not have an update hook
for the compute nodes).

Move installation of custom node package into specialized sub-recipes.

Create helper function is_custom_node? to detect whether a custom node
package is requested and trigger the custom node installer and update
recipes.

Adapt existing and added new InSpec tests to cover the custom node 
package installation both at AMI build time and at cluster creation.

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

Commit:
Move slurm recipes into phase specific folders

Signed-off-by: Enrico Usai <usai@amazon.com>

Commit:
Delete unused test_resource.rb file

This file is in the tests cookbook now.

Signed-off-by: Enrico Usai <usai@amazon.com>

Commit:
Move supervisord config into a separate recipe

Rename install/supervisord.rb to supervisord_install to avoid overlapping names.
Split ChefSpec tests.

Signed-off-by: Enrico Usai <usai@amazon.com>

Commit:
Pass dcv node attributes as variables to the template

In this way it's more clear which are the dependencies
and we avoid code duplication and typos.

Signed-off-by: Enrico Usai <usai@amazon.com>

dreambeyondorange committed:
Fix typo in CHANGELOG.md for CMA issue (#2310)

Commit:
Move creation of required directory as part of the config recipe

Signed-off-by: Enrico Usai <usai@amazon.com>

Commit:
Move supervisord restart into platform finalize recipe

Signed-off-by: Enrico Usai <usai@amazon.com>

Commit:
Remove scheduler plugin leftover

Signed-off-by: Enrico Usai <usai@amazon.com>

Commit:
Do not add dcv user if DCV is not installed

This cannot work on Ubuntu 20 ARM, on which DCV is not installed
because `imds-access.sh` script is unable to find the `dcv` user.

Signed-off-by: Enrico Usai <usai@amazon.com>

Commit:
Differentiate ptrace check according to the node type

ptrace_scope is configured to 0 in compute nodes and should be 1 in the head node.

Signed-off-by: Enrico Usai <usai@amazon.com>

Commit:
Check slurmd dependencies on graphic instances only

Signed-off-by: Enrico Usai <usai@amazon.com>

Commit:
Move sudoers and supervisord templates in the right folders

Signed-off-by: Enrico Usai <usai@amazon.com>

Commit:
Minor renaming of dcv parameter

Signed-off-by: Enrico Usai <usai@amazon.com>

Commit:
Add more details about code structure and testing in the README

Signed-off-by: Enrico Usai <usai@amazon.com>

Commit:
Minor cleanup to kitchen utility file

recipes-resources-validate file does not exist anymore.

Signed-off-by: Enrico Usai <usai@amazon.com>

Commit:
Move dcv authenticator virtualenv path in platform attributes

In this way we are sure that the variable is available for supervisord_config.rb

Signed-off-by: Enrico Usai <usai@amazon.com>

Commit:
Rename redhat_ubi to redhat_on_docker

UBI stands for Universal Base Image.
This is the Docker image created and maintained by Red Hat and updated regularly.

Rename internal variable to redhat_on_docker to avoid confusion.

https://hub.docker.com/r/redhat/ubi8
Signed-off-by: Enrico Usai <usai@amazon.com>

Commit:
Execute awsbatch cookbook Chefspec as GitHub actions

Signed-off-by: Enrico Usai <usai@amazon.com>

jdeamicis committed:
Create queue-nodelist mapping file for Slurm clusters

Create a JSON file with a mapping between all PC-managed queues
(partitions) and all PC-managed Slurm compute nodes.

Add unit test to cover the new logic for generating the partition-nodelist
specific mapping.

Add Changelog entry for the partition-nodelist mapping JSON creation.

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

dreambeyondorange committed:
Fix system test config for efa, remove duplicate test in intel_mpi, add an efa kitchen test suite for the head node, updated the efa test swtich for head and compute node (#2321)

jdeamicis committed:
Set autofix and indent 4 in pretty-format-json pre-commit hook.

Signed-off-by: Jacopo De Amicis <jdamicis@amazon.it>

dreambeyondorange committed:
[Ubuntu22] Add support for NVIDIA on Ubuntu 22 (#2322)

This change refactors the nvidia suites to the nvidia_ convention,
removes the duplicate nvidia_install suite, and verifies kitchen tests for Ubuntu 20 and 22,
refactors the uvm recipe to use `only_if` to force execution during the converge phase

Tests
- `nvidia_driver` install suite on Ubuntu 20 and 22
- `nvidia_dcgm` install and config suites on Ubuntu 20 and 22
- `nvidia_fabric_manager` install and config suites on Ubuntu 20 and 22
- `nvidia_cuda` install suite on Ubuntu 20 and 22
- `nvidia_gdrcopy` install and config suites on Ubuntu 20 and 22
- `nvidia_uvm` config suite on Ubuntu 20 and 22

Commit:
Relax Chef version requirements to use a more recent version of Chef/Cinc

This change permits to run tests locally with Chef 18
that is the version included in Chef Workstation 23.

Chef 17 is only compatible with Chef Workstation 22,
but this is not supported by MacBook with M1 chip.

Signed-off-by: Enrico Usai <usai@amazon.com>

Commit:
Fix raid_vol_array initialization

This must be an Array and not a string.

Code was failing with:
```
Chef::Exceptions::ValidationFailed: raid[add raid] (aws-parallelcluster-environment::update_shared_storages line 171) had an error: Chef::Exceptions::ValidationFailed: Property raid_vol_array must be one of: Array!  You passed "vol-0c79cc89dbb45d9cc,vol-00a87f00044e7df7f,vol-0431c0fbcd4d1bd33,vol-08662461b6a62b4ce,vol-01aab50f03443244f".
```

Signed-off-by: Enrico Usai <usai@amazon.com>

NSsirena committed:
Configure cluster user in LoginNodes

Signed-off-by: Nicola Sirena <nssirena@amazon.com>

NSsirena committed:
Mount EBS volumes in LoginNodes

Signed-off-by: Nicola Sirena <nssirena@amazon.com>

NSsirena committed:
Mount raid volumes in LoginNodes

Signed-off-by: Nicola Sirena <nssirena@amazon.com>

EddyMM committed:
Configure partitions to run jobs on nodes exclusively allocated based on the Cluster configuration parameter (JobExclusiveAllocation) (#2294)

Signed-off-by: Eddy Mwiti <eddmwiti@amazon.com>

NSsirena committed:
Create shared_login_nodes directory

Signed-off-by: Nicola Sirena <nssirena@amazon.com>

NSsirena committed:
Export shared_login_nodes directory

Also mounts /opt/intel in login nodes

Signed-off-by: Nicola Sirena <nssirena@amazon.com>

NSsirena committed:
Mount home and shared directories in login nodes

Also align the recipe to the rest of the cookbook
by using the internal resource `volume` for both
exporting and mounting shared FS.

Signed-off-by: Nicola Sirena <nssirena@amazon.com>

NSsirena committed:
Setup munge and mount /opt/slurm in login nodes

Signed-off-by: Nicola Sirena <nssirena@amazon.com>

NSsirena committed:
Extract mount /home as subrecipe

Signed-off-by: Nicola Sirena <nssirena@amazon.com>

NSsirena committed:
Block IDMS access to regular users on login nodes

Signed-off-by: Nicola Sirena <nssirena@amazon.com>

NSsirena committed:
Configure ActiveDirectory for LoginNodes

Also configure PAM to generate ssh keys

Signed-off-by: Nicola Sirena <nssirena@amazon.com>

Commit:
Add new utility `os-resources.py` to simplify creation of OS specific resources

The new utility `os-resources.py` script is able to copy the `resources/<name>/<name>_{src_platform}{src_version}`
into a `resources/<name>/<name>_{dst_platform}{dst_version}`, replacing the following patterns:
* `platform: '{src_platform}'`
* `platform?('{src_platform}')``
* `to_i == {src_version}``
* `platform_version: '{src_version}'`

This utility permits to automatically generate all the required files,
then they must be checked one by one to see if there are OS specific checks on them.
Templates and other specific files must be manually copied.

Signed-off-by: Enrico Usai <usai@amazon.com>

Commit:
Add test-os.sh utility to simplify run of kitchen tests for an OS

The script is able to search for all the test suites defined
into the functional cookbook and run them sequentially.

It must be executed from the root folder of the cookbook repo:
bash util/test-os.sh ec2 rhel8

The output will be saved into a `/tmp/kitchen-<os>-<date>` folder.

To run config tests is required to set a KITCHEN_${os}_AMI environment variable (e.g. KITCHEN_RHEL8_AMI)
or have a ParallelCluster AMI in the account.

Signed-off-by: Enrico Usai <usai@amazon.com>

Commit:
Export default ENV when PATH is required for command execution

exportfs and authconfig execution was failing with:
No such file or directory - exportfs/authconfig

### References
* `default_env true` documentation: https://docs.chef.io/resources/execute/#properties
* Related issue: https://github.com/sous-chefs/nfs/issues/106

Signed-off-by: Enrico Usai <usai@amazon.com>

Commit:
Export default ENV when PATH is required for command execution

mdadm execution during the update was failing with:
No such file or directory - mdadm

### References
* `default_env true` documentation: https://docs.chef.io/resources/execute/#properties
* Related issue: https://github.com/sous-chefs/nfs/issues/106

Signed-off-by: Enrico Usai <usai@amazon.com>

Commit:
Fix hooks execution

When the run-hooks script has been moved to a subdirectory we missed to fix the relative path.

Signed-off-by: Enrico Usai <usai@amazon.com>

Commit:
Improve kitchen tests hooks logging

Signed-off-by: Enrico Usai <usai@amazon.com>

Commit:
Skip deletion of resources when variables are not found in the env file

Signed-off-by: Enrico Usai <usai@amazon.com>

enrico-usai committed:
Replace deprecated (#2334)

`require_chef_omnibus` is going to be deprecated.
By using these new settings is possible in future to override some parameters,
for example we can specify a direct `download_url` if cinc package is not available
for a given product platform or version.

`product_name` is required when specifying `product_version`.

### Tests
Tested locally running some inspec tests.

### References
* https://docs.chef.io/workstation/config_yml_kitchen/#new-provisioner-settings

Signed-off-by: Enrico Usai <usai@amazon.com>

Commit:
Remove provisioner name

This is no longer required and causing issues:

```
>>>>>> Message: Could not load the 'cinc' provisioner from the load path. Did you mean: chef_apply, chef_base, chef_infra, chef_solo, chef_zero, chef_zero_capture, dokken, dummy, policyfile_zero, shell ? Please ensure that your provisioner is installed as a gem or included in your Gemfile if using Bundler.
```

### Tests
* Tested running local kitchen test.

### References
* Issue introduced with: https://github.com/aws/aws-parallelcluster-cookbook/pull/2334 on which I added the name just before pushing the patch, after the tests.

Signed-off-by: Enrico Usai <usai@amazon.com>

dreambeyondorange committed:
Skip the failing features for the Ubuntu 22 AMI build (#2333)

himani2411 committed:
Deprecate support for Ubuntu18 (#2330)

* Deprecation of Ubuntu1804

* Deprecation of Ubuntu1804 from tests

* Updating Metadata for cookbooks to remove Ubuntu18

---------

Co-authored-by: Himani Deshpande <himanidp@amazon.com>

Commit:
Split compute_base_spec into multiple controls and suites

Compute base was executing:
* 'aws-parallelcluster-platform::cluster_user'
* 'aws-parallelcluster-environment::raid'
* 'aws-parallelcluster-environment::shared_storages'
* 'aws-parallelcluster-environment::ebs'

so I splitted the check for the different phases on multiple tests.
I'm still keeping the mock.

### References
* https://github.com/aws/aws-parallelcluster-cookbook/pull/2290/commits/e7faebcc2ca9c26536e6cda20f391a5d7d281584

Signed-off-by: Enrico Usai <usai@amazon.com>

Commit:
Fix mock_compute shared storage recipes

During the export the only accepted value is "no_root_squash".

Removed useless node attribute parameters from the test.

Signed-off-by: Enrico Usai <usai@amazon.com>

Commit:
Split mock_compute_base.rb

Signed-off-by: Enrico Usai <usai@amazon.com>

NSsirena committed:
Add LoginNode role in CloudWatch agent configuration

It just add the new node-type.
The CW Agent configuration still doesn't work
on LoginNodes.

Signed-off-by: Nicola Sirena <nssirena@amazon.com>

NSsirena committed:
Define current and previous config path for LoginNodes

HeadNode retrieve the cluster config from S3 and shares it with
ComputeNodes using the "shared" folder and with LoginNodes using
the "shared_login_nodes" folder.

Signed-off-by: Nicola Sirena <nssirena@amazon.com>

NSsirena committed:
Ensure cluster config is retrieved only by the HeadNode

HeadNode retrieves the cluster config from S3 and shares it with
ComputeNodes using the "shared" folder and with LoginNodes using
the "shared_login_nodes" folder.

Signed-off-by: Nicola Sirena <nssirena@amazon.com>

NSsirena committed:
LoadClusterConfig take the config path as parameter

HeadNode and ComputeNodes read the config from the "shared" folder.
LoginNodes instead read it from "shared_login_nodes" folder.

Signed-off-by: Nicola Sirena <nssirena@amazon.com>

dreambeyondorange committed:
Skip the failing features for the Ubuntu 22 AMI build in kitchen tests as well (#2338)

Skip exit status for cloudwatch on Ubuntu 22

Commit:
Split resources and recipes tests

Move test suites in alphabetical order

Signed-off-by: Enrico Usai <usai@amazon.com>

lukeseawalker committed:
Avoid to start NFS server on compute nodes

Signed-off-by: Luca Carrogu <carrogu@amazon.com>

lukeseawalker committed:
Fix mock_compute shared storage recipes

During the export the only accepted value is "no_root_squash".

Removed useless node attribute parameters from the test.
Issue introduced with: #2336

Signed-off-by: Luca Carrogu <carrogu@amazon.com>

francesco-giordano committed:
Replace nvidia-persistenced with parallelcluster_nvidia service

`parallelcluster_nvidia` service ensures the creation of the block devices `/dev/nvidia0`
and it is needed by the `slurmd` service.

`parallelcluster_nvidia` starts the `nvidia-persistenced` or runs `nvidia-smi`
to avoid race condition with other services and avoids conflicts when using DLAMI with a gpu instance.

### Tests
* Modified ChefSpec to verify new changes.

### References
Backport of: https://github.com/aws/aws-parallelcluster-cookbook/pull/2341

Signed-off-by: Enrico Usai <usai@amazon.com>

Commit:
Improve hooks logging

Signed-off-by: Enrico Usai <usai@amazon.com>

Commit:
Create new mechanism to execute OS specific hooks

The kitchen tests will automatically execute hooks from kitchen/hooks folder
before executing the hooks for the specific test.

For example we can have a `kitchen/hooks/alinux2/pre_converge.sh` with the following content,
to automatically install some packages in the instance right after the startup, before Chef installation:

```
#!/bin/bash

if [ -n "${KITCHEN_INSTANCE_HOSTNAME}" ]; then
    export KITCHEN_EC2_USER='ec2-user'
    echo "Install libxcrypt-compat package by using SSH key: ${KITCHEN_SSH_KEY_PATH}"

    KITCHEN_EC2_INSTANCE_ID=$(ssh -o StrictHostKeyChecking=no -i "${KITCHEN_SSH_KEY_PATH}" \
      "${KITCHEN_EC2_USER}@${KITCHEN_INSTANCE_HOSTNAME}" 'sudo yum install -y some-package')
fi
```

This is useful for example when Chef cannot be installed in the AMI because due to a missing or
conflicting packages.

Signed-off-by: Enrico Usai <usai@amazon.com>

NSsirena committed:
Define helper method to identify LoginNodes

Recipes may have behaviors customized to the type of node where
they are executed. This change allows to test customized behaviors
for LoginNodes.

Signed-off-by: Nicola Sirena <nssirena@amazon.com>

NSsirena committed:
Test ebs volume configuration in LoginNodes

It does so by reusing and adapting the control used for compute nodes.

Signed-off-by: Nicola Sirena <nssirena@amazon.com>

Commit:
Remove PostgreSQL packages

The PostgreSQL packages have been added long time ago to support "PBSPro on RHEL based OSes".
Since PBSPro is not supported anymore we can safely remove the packages.

### References

https://github.com/aws/aws-parallelcluster-cookbook/commit/8d96814c7050b067c91dc3342d05710ff14c5cc9
Signed-off-by: Enrico Usai <usai@amazon.com>

NSsirena committed:
Test raid volume configuration in LoginNodes

It does so by reusing and adapting the control used for compute nodes.

Signed-off-by: Nicola Sirena <nssirena@amazon.com>

NSsirena committed:
Remove _netdev check from raid volumes test

The _netdev option is ignored in mount(8) by default.
It is not reported in /etc/mtab for raid volumes.

Signed-off-by: Nicola Sirena <nssirena@amazon.com>

NSsirena committed:
Test /opt/intel folder configuration in LoginNodes

It does so by reusing and adapting the control used for compute nodes.

Signed-off-by: Nicola Sirena <nssirena@amazon.com>

NSsirena committed:
Test shared volumes configuration in LoginNodes

It does so by reusing and adapting the control used for compute nodes.

Signed-off-by: Nicola Sirena <nssirena@amazon.com>

NSsirena committed:
Validate that /opt/slurm is mounted in Compute and Login nodes

The same subrecipe is executed for both node-types.
Mock the export of the shared folder.
Extract the subrecipe to mount the folder

Signed-off-by: Nicola Sirena <nssirena@amazon.com>

NSsirena committed:
fix cluster user spec

NSsirena committed:
Update changelog

Signed-off-by: Nicola Sirena <nssirena@amazon.com>

NSsirena committed:
Fix coostyle complaints

NSsirena committed:
Ensures compute fleet status is updated by HeadNode

Signed-off-by: Nicola Sirena <nssirena@amazon.com>

hanwen-pcluste committed:
Update CHANGELOG.md

EddyMM committed:
Configure partition to run jobs on nodes exclusively allocated based on the JobExclusiveAllocation parameter

Set [OverSubscribe/EXCLUSIVE](https://slurm.schedmd.com/slurm.conf.html#OPT_EXCLUSIVE) to partitions with JobExclusiveAllocation enabled

Signed-off-by: Eddy Mwiti <eddmwiti@amazon.com>

dreambeyondorange committed:
Update RSA keys to ED25519 on the cluster (#2350)

hanwen-cluster committed:
Increase EFS-utils watchdog poll interval to 10s

The default poll frequency is 1s, which could impact HPC workload performance

Signed-off-by: Hanwen <hanwenli@amazon.com>

dreambeyondorange committed:
Install libdbus-1-dev to build the slurm cgroups v2 plugin for Ubuntu 2204 (#2352)

hanwen-cluster committed:
Upgrade EFA to 1.24.0

Signed-off-by: Hanwen <hanwenli@amazon.com>

dreambeyondorange committed:
Update NVIDIA driver to account for what is available from Canonical (#2351)

NSsirena committed:
Export /opt/intel only when it exists.

On some OS/architecture is not available, so we need a guard.

Signed-off-by: Nicola Sirena <nssirena@amazon.com>

jdeamicis committed:
Split logrotate configuration into service-specific config files (#2354)

Create separate logrotate templates for separate log files/services rather
than one unique logrotate configuration file for all services.

This can be a first step towards migrating the logrotate configuration of
the various services to their respective recipes/resources/roles.

Create helper function to generate a list of logrotate config files within a recipe.

Modify ChefSpec test for a split logrotate config.

lukeseawalker committed:
Fix raise when no dynamoDB entry is found

When dynamoDB entry isn't yet present, the command output is an empty string

Signed-off-by: Luca Carrogu <carrogu@amazon.com>

lukeseawalker committed:
Add retries when no dynamoDB entry is found

Add retries when no dynamoDB entry is found, for a total wait time of 2 additional minutes.

Signed-off-by: Luca Carrogu <carrogu@amazon.com>

Commit:
Move Intel HPC configuration after setup of shared folder

The config step of the Intel HPC installer installs rpms from `/opt/intel/rpms/`.
This folder must be mounted before executing intel hpc installation steps.

Without this patch the installation works for the head node but
fails for the compute nodes because unable to see the rpms in the shared folder.

Signed-off-by: Enrico Usai <usai@amazon.com>

dreambeyondorange committed:
Update NFS configure to stop the service on compute nodes and the test to check that the service is stopped (#2362)

rkilpadi committed:
Add basic Spack installation resource

dreambeyondorange commented on cookbooks/aws-parallelcluster-platform/resources/spack/partial/_spack_common.rb:
```diff
@@ -0,0 +1,98 @@
+         default: node['cluster']['cluster_user']
+
+property :spack_root, String, required: false,
+         default: "/home/#{node['cluster']['cluster_user']}/spack"
```
let's define an env attribute for this in cookbooks/aws-parallelcluster-environment/attributes

dreambeyondorange commented on cookbooks/aws-parallelcluster-platform/resources/spack/partial/_spack_common.rb:
```diff
@@ -0,0 +1,98 @@
+  template '/etc/profile.d/spack.sh' do
+    cookbook 'aws-parallelcluster-platform'
+    source 'spack/spack.sh.erb'
+    owner 'root'
```
could this be owned by the user?

dreambeyondorange commented on cookbooks/aws-parallelcluster-platform/resources/spack/spack_amazon2.rb:
```diff
@@ -0,0 +1,27 @@
+end
+
+def libfabric_path
+  '/opt/amazon/efa/lib64/pkgconfig/libfabric.pc'
```
this should be an env attribute as well.

dreambeyondorange commented on cookbooks/aws-parallelcluster-platform/resources/spack/partial/_spack_common.rb:
```diff
@@ -0,0 +1,98 @@
+
+  # Pull architecture dependent package config
+  begin
+    template "#{spack_configs_dir}/packages.yaml" do
```
This is where we start to add value!

dreambeyondorange commented on cookbooks/aws-parallelcluster-platform/resources/spack/partial/_spack_common.rb:
```diff
@@ -0,0 +1,98 @@
+    user new_resource.spack_user
+    environment({ "HOME" => "/home/#{new_resource.spack_user}" })
+    code <<-SPACK
+      source #{new_resource.spack_root}/share/spack/setup-env.sh
```
For those unfamiliar with Spack, can you add one line comments to these to briefly explain what they're doing?

dreambeyondorange commented on cookbooks/aws-parallelcluster-platform/resources/spack/spack_ubuntu.rb:
```diff
@@ -0,0 +1,28 @@
+# frozen_string_literal: true
```
This shouldn't be just plain ubuntu.  We have deprecated ubuntu 18 so this is really just ubuntu 20

dreambeyondorange commented on cookbooks/aws-parallelcluster-platform/test/controls/spack_spec.rb:
```diff
@@ -0,0 +1,66 @@
+spack = "#{node['cluster']['spack']['root']}/bin/spack"
+
+control 'tag:install_spack_correctly_installed' do
+  only_if { !os_properties.on_docker? }
```
Why can't this work on docker?

dreambeyondorange commented on cookbooks/aws-parallelcluster-platform/test/controls/spack_spec.rb:
```diff
@@ -0,0 +1,66 @@
+end
+
+control 'tag:install_spack_can_install_packages' do
+  only_if { !os_properties.on_docker? }
```
same, why doesn't this work on docker?
